controlplane/kubeadm/internal/etcd_client_generator.go

```diff
@@ -72,26 +79,54 @@ func (c *EtcdClientGenerator) forFirstAvailableNode(ctx context.Context, nodeNam
 
 // forLeader takes a list of nodes and returns a client to the leader node.
 func (c *EtcdClientGenerator) forLeader(ctx context.Context, nodeNames []string) (*etcd.Client, error) {
-	var errs []error
+	// This is an additional safeguard for avoiding this func to return nil, nil.
+	if len(nodeNames) == 0 {
+		return nil, errors.New("invalid argument: forLeader can't be called with an empty list of nodes")
+	}
 
+	nodes := sets.NewString()
+	for _, n := range nodeNames {
+		nodes.Insert(n)
+	}
+
+	// Loop through the existing control plane nodes.
+	var errs []error
 	for _, nodeName := range nodeNames {
+		// Get a temporary client to the etcd instance hosted on the node.
 		client, err := c.forFirstAvailableNode(ctx, []string{nodeName})
 		if err != nil {
 			errs = append(errs, err)
 			continue
 		}
 		defer client.Close()
+
+		// Get the list of members.
 		members, err := client.Members(ctx)
 		if err != nil {
 			errs = append(errs, err)
 			continue
 		}
+
+		// Get the leader member.
+		var leaderMember *etcd.Member
 		for _, member := range members {
-			if member.Name == nodeName && member.ID == client.LeaderID {
-				return c.forFirstAvailableNode(ctx, []string{nodeName})
+			if member.ID == client.LeaderID {
+				leaderMember = member
 			}
 		}
-	}
 
+		// If we found the leader and it is one of the nodes,
+		// get a connection to the etcd leader via the node hosting it.
+		if leaderMember != nil && nodes.Has(leaderMember.Name) {
+			return c.forFirstAvailableNode(ctx, []string{leaderMember.Name})
+		}
+
+		// If it is not possible to get a connection to the leader via existing nodes,
+		// it means that the control plane is an invalid state, with an etcd member - the current leader -
+		// without a corresponding node.
+		// TODO: In future we can eventually try to automatically remediate this condition by moving the leader
+		//  to another member with a corresponding node.
+		return nil, errors.Errorf("could not find the node corresponding to the current etcd leader %x", client.LeaderID)
```

Comment:
If a node in the list that we can connect to doesn't for whatever reason have the leaderID, shouldn't we try the next one? From my understanding this for loop only tries to connect to the first node in the list and if we can't find the leader, it always errors out.

Reply:
> If a node in the list that we can connect to doesn't for whatever reason have the leaderID

This is means that we have an etcd member in weird state, and this kind of means that we have etcd members with different opinions about who the leader is.
I don't think that KCP should move on and take decisions in this case...